20240927-multiple-object-deletion.md

```diff
@@ -0,0 +1,163 @@
+tags: [ "gateway", "satellite", "satellite/metainfo", "uplink" ]
+---
+
+# Multiple Object Deletion
+
+## Essentials
+
+### Header
+
+Date: 2024-09-27
+
+Owner: Jeremy Wharton
+
+Accountable:
+- Edge team
+
+Consulted:
+- Product team
+
+Informed:
+- QA team
+
+### Context
+
+The satellite metainfo RPC service provides an API through which objects stored on the Storj network may be interacted with. While this service exposes an endpoint through which individual objects may be deleted, it lacks an endpoint for deleting multiple objects in a single request. This limitation impacts services that depend on the metainfo RPC service, such as our S3-compatible gateway, by requiring separate requests for each object in a bulk deletion operation.
+
+This approach has several issues:
+
+- Network overhead: The amount of time spent transmitting and processing network requests increases with the number of objects to be deleted. This not only delays bulk operations but also consumes unnecessary bandwidth.
+
+- Rate limiting: Each deletion request counts against the deletion rate limit of the project associated with the request. This reduces the amount of deletion operations that can be performed within a given time frame. Additionally, this may cause bulk deletions to fail partway if the rate limit is reached mid-operation.
+
+- Database overhead: Each deletion involves the execution of a separate SQL query, increasing the load on the database as the object count grows. Implementing bulk deletions gives us the opportunity to batch queries or delete multiple objects in a single query, reducing database overhead.
+
+For further context, the gateway currently supports Amazon S3's [DeleteObjects](https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteObjects.html) operation, which allows for the deletion of multiple objects with a single HTTP request. However, the lack of native support for bulk deletions in the metainfo RPC service makes this operation subject to the aforementioned issues, as the gateway must translate every S3 request to a series of RPC requests.
+
+### Goals
+
+The goal of the implementation described by this document is to allow multiple objects to be deleted in a single request to the satellite.
+
+### Approach / Design
+
+#### Satellite
+
+The satellite metainfo RPC service must be extended with an endpoint allowing for the deletion of multiple objects in a single request.
+
+Requests directed to this endpoint must contain the following information:
+
+- The requester's API key
+- The bucket from which objects will be deleted
+- An indication of whether governance mode Object Lock restrictions should be bypassed
+- An indication of whether quiet mode should be enabled.
+  - With quiet mode enabled, only errors are returned; information about successful deletions is omitted from responses. Supporting this mode would allow us to save bandwidth on clients that use it.
+- A list of objects to delete
```

Comment:
It might be worth mentioning the 1000 limit up-front here, since that's a design decision influenced by S3. I noticed you already had it in the test plan.